sourcegraph/cody
# Autocomplete: Improve sampling code and prepare for Honeycomb export

This PR improves the client site traces for autocomplete and makes it ready to be propagated to Honeycomb. An example trace can be found here: https://ui.honeycomb.io/sourcegraph/environments/cody/datasets/cody-client/result/z87pN9yLSan/trace/rv3ajXLDNV7?fields[]=s_name&fields[]=s_serviceName&span=2edb453047488acc

<img width="1673" alt="Screenshot 2024-02-07 at 18 16 09" src="https://github.com/sourcegraph/cody/assets/458591/89abf1fe-0394-4c84-9912-e7141a365a2b">

The changes can be summarized as:
- We add a suggested event that we set when a completion callback makes it to the end and will result in being visible by a user
- We add all exposed feature flags after the completion is ready (this means it will include eventual flags that are being evaluated as part of the completion request) as well as all the metadata that would also be added to analytics pings
- Fix a bug that caused spans to be much shorter than they were. Whoopsie!

These changes were tested with the following OTel collector processor config:

```
processors:
  tail_sampling:
    decision_wait: 30s
    num_traces: 100000
    expected_new_traces_per_sec: 10
    policies:
      name: and
      type: and
      and:
        and_sub_policy:
          # We only want to sample traces for completions that are displayed to the user. We track
          # this with a suggested event on the root span
          - name: completion-was-suggested
            type: ottl_condition
            ottl_condition:
              error_mode: ignore
              spanevent:
                - "name == \"suggested\""
          # Only track events that originate from the completion provider
          - name: is-completion-event
            type: ottl_condition
            ottl_condition:
              error_mode: ignore
              span:
                - "name == \"autocomplete.provideInlineCompletionItems\""
          # Only track spans for users that are expose to the tracing experiment
          - name: is-tracing-enabled
            type: ottl_condition
            ottl_condition:
              error_mode: ignore
              span:
                - "attributes[\"cody-autocomplete-tracing\"] == true"
          # # Ignore completion callbacks that return the last candidate. We only care for traces that
          # # show a new completion
          - name: is-completion-event
            type: ottl_condition
            ottl_condition:
              error_mode: ignore
              spanevent:
                - "attributes[\"source\"] != \"LastCandidate\""
```

## Test plan

- Configure cody locally and run otel configurator (see https://docs.sourcegraph.com/dev/how-to/opentelemetry_local_dev)
- Apply the filter that I pasted above into the honeycomb config and make sure that it's piped correctly 
   - **UPDATE:** Use the latest version in the comment below
- In `OpenTelemetryService.node.ts` remove the dev mode `spanProcessor`.
- Look the events pipe into honeycomb (e.g.  https://ui.honeycomb.io/sourcegraph/environments/cody/datasets/cody-client/result/z87pN9yLSan/trace/rv3ajXLDNV7?fields[]=s_name&fields[]=s_serviceName&span=2edb453047488acc)
